src/VolumeCache.ts

```diff
@@ -0,0 +1,306 @@
+type XYZ<T> = { x: T; y: T; z: T };
+export type VolumeScaleDims = XYZ<number>;
+type CacheEntryExtent = XYZ<[number, number]>;
+
+// The following two very similar types are kept separate because we may later want to allow more
+// complex queries with respect to scale, e.g. "get the largest available scale within this range"
+export type DataArrayExtent = CacheEntryExtent & {
+  scale: number;
+  time: number;
+  channel: number;
+};
+
+export type QueryExtent = CacheEntryExtent & {
+  scale: number;
+  time: number;
+};
+
+type CacheEntry = {
+  /** The data contained in this entry */
+  // TODO allow more types of `TypedArray` to be stored together in the cache?
+  data: Uint8Array;
+  /** The subset of the volume covered by this entry */
+  size: CacheEntryExtent;
+  /** The previous entry in the LRU list */
+  prev: MaybeCacheEntry;
+  /** The next entry in the LRU list */
+  next: MaybeCacheEntry;
+  /** The array which contains this entry */
+  parentArr: CacheEntry[];
+};
+type MaybeCacheEntry = CacheEntry | null;
+
+type VolumeCacheScale = {
+  // Entries are indexed by T and C, then stored in lists of ZYX subsets
+  data: CacheEntry[][][];
+  size: VolumeScaleDims;
+};
+
+type CachedVolume = {
+  scales: VolumeCacheScale[];
+  times: number;
+  channels: number;
+};
+
+const extentIsEqual = (a: CacheEntryExtent, b: CacheEntryExtent): boolean => {
+  return (
+    a.x[0] === b.x[0] &&
+    a.x[1] === b.x[1] &&
+    a.y[0] === b.y[0] &&
+    a.y[1] === b.y[1] &&
+    a.z[0] === b.z[0] &&
+    a.z[1] === b.z[1]
+  );
+};
+
+// Extent ranges are inclusive on both ends
+const dimSize = ([min, max]: [number, number]): number => max + 1 - min;
+const extentVolume = ({ x, y, z }: CacheEntryExtent): number => dimSize(x) * dimSize(y) * dimSize(z);
+const dimInvalid = ([min, max]: [number, number]): boolean => min > max;
+const extentInvalid = ({ x, y, z }: CacheEntryExtent): boolean => dimInvalid(x) || dimInvalid(y) || dimInvalid(z);
+
+export default class VolumeCache {
+  private store: CachedVolume[];
+  public readonly maxSize: number;
+  private currentSize: number;
+  private currentEntries: number;
+
+  // Ends of a linked list of entries, to track LRU and evict efficiently
+  private first: MaybeCacheEntry;
+  private last: MaybeCacheEntry;
```

Comment:
This is a solid point! It might actually make the most sense to flip `prev` and `next` in `CacheEntry` - I was thinking "added previously/next," but you're right that the convention is "previous/next in the list."

Reply:
Thank you for making this change, it makes your code a lot more readable!